compiler/src/iree/compiler/InputConversion/MHLO/ConvertMHLOToLinalgExt.cpp

```diff
@@ -470,6 +471,121 @@ struct ReverseOpConversion : public OpConversionPattern<mhlo::ReverseOp> {
+    uint64_t kDim = inputValuesType.getRank() - 1;
+
+    // Create a range of indicies for the input tensor match input shape but
+    // with i32 type.
+    SmallVector<Value> dynSizes;
+    for (auto en : llvm::enumerate(inputValuesType.getShape())) {
+      if (en.value() == ShapedType::kDynamicSize) {
+        dynSizes.push_back(
+            rewriter.create<tensor::DimOp>(loc, adaptor.operand(), en.index()));
+      }
+    }
+
+    Value initAcc = rewriter.create<linalg::InitTensorOp>(
+        loc, dynSizes, inputValuesType.getShape(), indicesElementType);
+
+    // Create the indexing maps for the generic
+    size_t inputRank = inputValuesType.getRank();
+    AffineMap mapIdentity = rewriter.getMultiDimIdentityMap(inputRank);
+    SmallVector<AffineMap> indexingMaps = {mapIdentity};
+    SmallVector<StringRef> iterators(inputRank, "parallel");
+    Value inputIndices =
+        rewriter
+            .create<linalg::GenericOp>(
+                loc, /*resultTensorTypes=*/initAcc.getType(),
+                /*inputs=*/ValueRange{}, /*outputs=*/ValueRange{initAcc},
+                indexingMaps, iterators,
+                [=](OpBuilder &b, Location loc, ValueRange args) {
+                  // Indicies increase along the k dimension
+                  Value index = b.create<linalg::IndexOp>(loc, kDim);
+                  Value intIndex = b.create<arith::IndexCastOp>(
+                      loc, indicesElementType, index);
+                  b.create<linalg::YieldOp>(loc, intIndex);
+                })
+            .getResult(0);
+
+    // Create and initialize output tensors for LinalgExt TopK results
+    // Define the output types based on the results of CHLO TopK
+    Value initTensorOutputValues = rewriter.create<mlir::linalg::InitTensorOp>(
+        loc, dynSizes, outputValuesType.getShape(), valueElementType);
+    Value initTensorOutputIndices = rewriter.create<mlir::linalg::InitTensorOp>(
+        loc, dynSizes, outputIndicesType.getShape(), indicesElementType);
+    // Initialize indices to 0 and values to negative infinity
+    Attribute negInfAttr;
+    if (auto intType = valueElementType.dyn_cast<IntegerType>()) {
+      negInfAttr = rewriter.getIntegerAttr(
+          intType, APInt::getSignedMinValue(intType.getWidth()));
+    } else {
+      auto negApFloat = APFloat::getInf(
+          valueElementType.cast<FloatType>().getFloatSemantics(),
+          /*Negative=*/true);
+      negInfAttr = rewriter.getFloatAttr(valueElementType, negApFloat);
+    }
+    Value negInf = rewriter.create<arith::ConstantOp>(loc, negInfAttr);
+    Value zero = rewriter.create<arith::ConstantOp>(
+        loc, rewriter.getZeroAttr(indicesElementType));
+    Value negInfTensor =
+        rewriter.create<linalg::FillOp>(loc, negInf, initTensorOutputValues)
+            .result();
+    Value zeroTensor =
+        rewriter.create<linalg::FillOp>(loc, zero, initTensorOutputIndices)
+            .result();
```

Comment:
I still think that we don't have to initialize output indices...

I think the assumption is that input values are all non-negative. In this context, initializing output values is enough. Because top_k compares the values firstly. The indices are considered only when values are equal.

Without the assumption, this is still wrong. Say that if all the inputs are negative. We'll return K `(neg ,0)` pairs?

Reply:
to clarify, we're not making any assumptions about the input. Values can be negative or positive. The choice of `negInf` is to ensure any value higher than the lowest possible value is chosen, and if all the values happen to also be `negInf` then they'll fall to the second case where the "first" neg infinity is chosen. 

Actually, now that you point this out, I realize that the `zeroTensor` assigned to the indices is actually incorrect; it should be `maxInf` instead so that if all the values are negative infinity, the first occurrences are chosen. I've updated the code to reflect this. 

Ultimately, as stated above, I believe both these values do need to be initialized. Also as Mahesh stated, I do believe if the values aren't initialized, they are essentially undefined, which will lean do undefined behavior with these comparisons.

Reply:
Although I'm not really convinced that the initialization is needed, I think it shouldn't be a blocker for landing the PR.

The topk operation returns the indices. Since the ivs and k are known, we should be able to check what values to use. It can be checked by comparing ivs and k to me. Maybe the current implementation puts some limits on it.